apps/astarte_appengine_api/test/support_v2/helpers/database.ex

```diff
@@ -0,0 +1,622 @@
+    binaryblobarray_value list<blob>,
+    datetimearray_value list<timestamp>,
+
+    PRIMARY KEY((device_id, interface_id, endpoint_id, path), value_timestamp, reception_timestamp, reception_timestamp_submillis)
+  )
+  """
+
+  @create_groups_table """
+  CREATE TABLE :keyspace.grouped_devices (
+    group_name varchar,
+    insertion_uuid timeuuid,
+    device_id uuid,
+    PRIMARY KEY ((group_name), insertion_uuid, device_id)
+  )
+  """
+
+  ###
+  ### Insert
+  @insert_interface """
+  INSERT INTO :keyspace.interfaces (
+    name,
+    interface_id,
+    major_version,
+    minor_version,
+    type,
+    automaton_accepting_states,
+    automaton_transitions,
+    aggregation,
+    ownership,
+    storage,
+    storage_type
+    ) VALUES (
+      :name,
+      :interface_id,
+      :major_version,
+      :minor_version,
+      :type,
+      :automaton_accepting_states,
+      :automaton_transitions,
+      :aggregation,
+      :ownership,
+      :storage,
+      :storage_type
+    )
+  """
+
+  @insert_device """
+  INSERT INTO :keyspace.devices
+  (
+     device_id,
+     aliases,
+     attributes,
+     connected,
+     last_connection,
+     last_disconnection,
+     first_registration,
+     first_credentials_request,
+     last_seen_ip,
+     last_credentials_request_ip,
+     total_received_msgs,
+     total_received_bytes,
+     inhibit_credentials_request,
+     introspection,
+     introspection_minor,
+     exchanged_msgs_by_interface,
+     exchanged_bytes_by_interface
+  )
+  VALUES
+  (
+    :device_id,
+    :aliases,
+    :attributes,
+    :connected,
+    :last_connection,
+    :last_disconnection,
+    :first_registration,
+    :first_credentials_request,
+    :last_seen_ip,
+    :last_credentials_request_ip,
+    :total_received_msgs,
+    :total_received_bytes,
+    :inhibit_credentials_request,
+    :introspection,
+    :introspection_minor,
+    :exchanged_msgs_by_interface,
+    :exchanged_bytes_by_interface
+  )
+  """
+
+  @insert_alias """
+  INSERT INTO :keyspace.names (
+    object_name,
+    object_type,
+    object_uuid
+  )
+  VALUES (
+    :object_name,
+    1,
+    :object_uuid
+  )
+  """
+
+  @insert_group """
+  INSERT INTO :keyspace.grouped_devices (
+    group_name,
+    insertion_uuid,
+    device_id
+    )
+  VALUES
+  (
+    :group_name,
+    :insertion_uuid,
+    :device_id
+  )
+  """
+
+  @insert_pubkeypem """
+  INSERT INTO :keyspace.kv_store (group, key, value)
+  VALUES ('auth', 'jwt_public_key_pem', varcharAsBlob(:pem));
+  """
+
+  ###
+  ### Update
+  @update_device """
+  UPDATE :keyspace.devices
+  SET :field = :field + :data
+  WHERE device_id = :device_id
+  """
+
+  ###
+  ### Select
+  @select_interfaces """
+  SELECT * FROM :keyspace.interfaces WHERE name IN :names
+  """
+
+  @select_devices """
+  SELECT * FROM :keyspace.devices WHERE device_id IN :device_ids
+  """
+
+  ###
+  ### Delete
+  @delete_interface """
+  DELETE FROM :keyspace.interfaces WHERE name = :name
```

Comment:
If tests from different cases run in different keyspaces and, inside a case, each test populates the tables it needs and truncates them on exit, you get database isolation for free, without the need to write more complex queries for deleting only some data.
A quick sketch of a possible way of handling it:
```elixir
defmodule MyTest do
  use ExUnit.Case
  
  setup_all do
     create_keyspace!("keyspace_#{System.unique_integer()}")
  end
  
  setup do
    # get keyspace from context 
    insert!(keyspace, devices_fixture())
    on_exit(fn -> run!("TRUNCATE TABLE devices") end)
  end

  test "my test 1" do
  # use the `devices` table
  end
end
```

Reply:
I have already discussed this with @rbino. The point is to write as little code as possible in each test file. Cases are made to achieve this